### ISSUE TITLE
Integration: start a CI task upon receiving `push` webhook

### ISSUE THREAD

**Issue by nolanderc** [KTH-Software-Engineering-DD2480/ci#32]
After receiving a `push` webhook, we want to:

1. Start a new thread (so that the HTTP server is not blocked on the build process)
2. Change the GitHub commit status to "in progress"
3. Clone the repository to a new folder and checkout the given commit.
4. Run `gradle build`
5. Run `gradle test`
6. Send the output/results to the persistence layer (this could probably be done in a streaming fashion, if we want)
7. Change the GitHub commit status to either "failure" or "success"
8. Exit the thread

Note that we should probably assign each task its own unique identifier (such as a monotonically increasing integer) in case multiple webhooks come in simultaneously on different commits (for example, each task gets its own directory, inside which it clones the repo (probably in a subfolder, so we can safely delete it) and stores the build output). If we reuse the same folder for the repo, the two CI tasks might clash. Alternatively, we could implement some sort of job-queue to prevent multiple tasks from executing in parallel.

**Comment by mantaur** [comment]
I don't really follow the part about assigning each task its own unique identifier, but the thoughts on handling concurrent commits and in turn webhooks seems sound. At first thought the folder separation sounded like a good idea but I imagine there's built in concurrency handling in our HTTP server run by jetty which could help us use the job-queue system instead.

**Comment by nolanderc** [comment]
I see what you are saying, but the problem is that we need to respond to the GitHub webhook (ie. return from the HTTP handler) as soon as possible. We shouldn't hold the connection open for longer than necessary. So spinning up a new thread for each task will be required.

About the unique identifiers: we basically just need some way to distinguish two different CI tasks/commits from each other. And using this identifier to name different folders for each task just felt like the simplest way of doing things. 

The job queue idea is separate, and would essentially just boil down to us having a single thread executing all the tasks sequentially. That is, a job would come in, and get queued, then the task-thread works through them in a FIFO order. We would still have the same issues around keeping the old log files separate, but we don't have to worry about conflicts in the main repo (at the cost of only being able to run jobs sequentially, but that will likely not be a problem for this assignment).

**Comment by psalqvist** [comment]
Just one thing that I'm wondering about to make sure I understand. If we use a single thread to run the FIFO queue. Do we still need a unique identifier for the folders? Couldn't we just clone the repo, execute 2-7 and then delete the repo. If we run it sequentially we won't need to name folders uniquely, right?

**Comment by nolanderc** [comment]
Correct!

**Comment by psalqvist** [comment]
Great. Furthermore, I have created a class ExecuteJob that should run a function execute (performs 2-7), which is going to be called by the thread that runs through our job queue. Previously we had a discussion regarding that exceptions in updateStatus should be propagated up the call stack to be handled by the caller. Does this apply here as well? Do we want to propagate the exceptions up to the thread for them to be handled there, or should we handle exceptions in the ExecuteJob class?

**Comment by psalqvist** [comment]
@mantaur you are implementing the queue right? What do you think?

**Comment by nolanderc** [comment]
@psalqvist Conceptually, the "job queue" is just a loop:
```
while (server_is_running) {
    item = next_work_item(); // <- the queue @mantaur is working on
    set_status(IN_PROGRESS);
    status = execute(item);
    set_status(status);
}
```
The best place to handle any exceptions is probably in the body of that loop (around the `execute` and `set_status` functions). There's not really much we can do about them, than just to add them to the persistent log files.

**Comment by psalqvist** [comment]
Alright, so as for now I just let the exceptions propagate from execute to whoever is calling the function? And then we just handle the exceptions in the loop? Right now I'm running both set_status(IN_PROGRESS) and set_status(status) in the function called execute. Would this make things less readable or is it ok?

**Comment by nolanderc** [comment]
That's probably OK, but we also have to consider the case when there is an exception. In this case we want to set the GitHub commit status to `ERROR`, so having all these functions in the same place might be more obvious to the reader. Alternatively, you could just add the `try/catch` to the `execute` function (ie. keep the loop simple, and move all logic into the execute function).

**Comment by psalqvist** [comment]
I'll add try/catch to the execute function. Whenever we catch an exception, I'll just set the commit status to error, and save the exception in the persistent log files?

**Comment by nolanderc** [comment]
Yes! Currently the `output` of the procedure is named `gradle_output`. Though I would like if we instead could make use of multiple different "stages". For example, we would most likely want to separate any output from the clone, build and tests steps, preferably. But for now, just put everything into a single string and send it to the log files.

**Comment by nolanderc** [comment]
fixed in #44
